src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rest/QueryBatchRequestHandler.java

```diff
@@ -0,0 +1,309 @@
+              DEFAULT_MAX_DATAPOINTS);
+      sendResponse(exchange, queryResponse, HttpURLConnection.HTTP_OK);
+    } catch (InvalidParameterException e) {
+      LOG.error(
+              (Supplier<?>)
+                      () ->
+                              new ParameterizedMessage(
+                                      "QueryException {} ExceptionCode: {}.",
+                                      e.toString(),
+                                      StatExceptionCode.REQUEST_ERROR.toString()),
+              e);
+      StatsCollector.instance().logException(StatExceptionCode.REQUEST_ERROR);
+      String response = "{\"error\":\"" + e.getMessage() + ".\"}";
+      sendResponse(exchange, response, HttpURLConnection.HTTP_BAD_REQUEST);
+    } catch (Exception e) {
+      LOG.error(
+              (Supplier<?>)
+                      () ->
+                              new ParameterizedMessage(
+                                      "QueryException {} ExceptionCode: {}.",
+                                      e.toString(),
+                                      StatExceptionCode.REQUEST_ERROR.toString()),
+              e);
+      StatsCollector.instance().logException(StatExceptionCode.REQUEST_ERROR);
+      String response = "{\"error\":\"" + e.toString() + "\"}";
+      sendResponse(exchange, response, HttpURLConnection.HTTP_INTERNAL_ERROR);
+    }
+  }
+
+  @VisibleForTesting
+  public int appendMetrics(Long timestamp, List<String> metrics, StringBuilder builder, int maxDatapoints) throws Exception {
+    maxDatapoints += 1;
+    builder.append("\"");
+    builder.append(timestamp);
+    builder.append("\":{");
+    MetricsDB db = MetricsDB.fetchExisting(timestamp);
+    for (int metricIndex = 0, numMetrics = metrics.size(); metricIndex < numMetrics; metricIndex++) {
+      String metric = metrics.get(metricIndex);
+      Result<Record> results = db.queryMetric(metric, MetricsModel.ALL_METRICS.get(metric).dimensionNames, maxDatapoints);
+      if (results != null) {
+        maxDatapoints -= results.size();
+        if (maxDatapoints == 0) {
+          throw new InvalidParameterException(String.format("requested data exceeds the %d datapoints limit", DEFAULT_MAX_DATAPOINTS));
+        }
+        builder.append("\"");
+        builder.append(metric);
+        builder.append("\":");
+        builder.append(results.formatJSON());
+        for (metricIndex += 1; metricIndex < numMetrics; metricIndex++) {
+          metric = metrics.get(metricIndex);
+          results = db.queryMetric(metric, MetricsModel.ALL_METRICS.get(metric).dimensionNames, maxDatapoints);
+          if (results != null) {
+            maxDatapoints -= results.size();
+            if (maxDatapoints == 0) {
+              throw new InvalidParameterException(String.format("requested data exceeds the %d datapoints limit", DEFAULT_MAX_DATAPOINTS));
+            }
+            builder.append(",\"");
+            builder.append(metric);
```

Comment:
I've already commented on this to Ricardo. I think it's better to avoid tight coupling between the JSON response construction and the actual sampling and querying of the metrics. 

However, since today there are no components that would actually need the batching logic outside of the REST layer, and since this is an L0 solution, I've said that I'd note my concern here and +1.

Reply:
Does anyone else have a strong opinion on this? @yojs @vigyasharma

Reply:
If we find a use for such decoupling, perhaps we can modify the implementation with such changes. However, for now, this approach will work, and it removes the need for additional investigation (for example, this would increase processing time, and stringifying a Batch Metrics Response object would double the peak memory usage of this component).

Reply:
In no case we should be manually constructing json. Please use the Gson library. It will make it easier and less error prone. Maybe you can look at HotNodeSummary::toJson()

Reply:
I agree with joydeep and sid. There are standard tested libraries that we should leverage. Can move to a separate PR as a follow up.

Reply:
Sure, will raise a new enhancement issue about it #416